Feat(code): Add Intrasolar ship category & update flight check

MadisonianX opened the pull request:
## Issue
When creating an alternate Intrasolar start in a scaled up Sol with no hyperdrive, every attempted departure triggers the flight check warning for no hyperdrive. 

## Summary

This adds a simple `Intrasolar` ship category and an appropriate change to the flight check. Any plugin can use the category to exempt content from the hyperdrive/jump drive flight check warnings for true intrasolar starts. Doesn't affect the flight check for any other ship category.

## Video
Before:
https://github.com/user-attachments/assets/3b21531a-6895-4154-9a48-1a3486ef91e2

After:
https://github.com/user-attachments/assets/8638adc8-1adb-4395-bbc1-cb728cf3e7e8


## Usage examples
```
ship "Used Shuttle"
	attributes
		category "Intrasolar"
```
## Testing Done
Created a Used Shuttle with the `Intrasolar` category and no hyperdrive. Tried to depart without the flight check code change, received the warning. 

With the flight check code change, departed fine with no warning. Purchased ships from any other category and sold the hyperdrive, warning is still there when trying to depart.

TheGiraffe3 reviewed (approved):
I don't suppose ES main would be interested in getting this?
Although I suppose you don't have a fork or branch without Delta's extraneous commits...

MadisonianX commented:
> I don't suppose ES main would be interested in getting this? Although I suppose you don't have a fork or branch without Delta's extraneous commits...

Feel free to add anything to the main branch, I'm still trying to get my head around how branches and forks all work so I'm submitting stuff here for a more stress-free environment than that crowd. Once I get more comfortable with these processes I'll probably start submitting content there too.

TheGiraffe3 commented:
Actually, how easy would it be to make this an attribute instead of a category?
Say I want to have a fighter that doesn't produce warnings. If I give it the "Intrasolar" category nothing can carry it; if I give it the "Fighter" category it'll produce warnings.
Creating it with the "Fighter" category and giving it an "intrasolar" attribute might be easier.

MadisonianX commented:
> Actually, how easy would it be to make this an attribute instead of a category? Say I want to have a fighter that doesn't produce warnings. If I give it the "Intrasolar" category nothing can carry it; if I give it the "Fighter" category it'll produce warnings. Creating it with the "Fighter" category and giving it an "intrasolar" attribute might be easier.

It can easily be added as an attribute as well, the problem is that it will exempt the Fighter from the hyperdrive flight check but not the insufficient bays flight check from PlayerInfo, so there'd still be a different warning every launch unless you tweak or remove the "no bays?" check too. 

Tested with this code in Ship::FlightCheck():

`	double intrasolar = attributes.Get("intrasolar");
`

and then in the `if(checks.empty()) ` block:

```
		if(canBeCarried && !intrasolar)
		{
			if(!hyperDrive && !jumpDrive)
				checks.emplace_back("no hyperdrive?");
			if(fuelCapacity < navigation.JumpFuel())
				checks.emplace_back("no fuel?");
		}
		if(!canBeCarried)
		{
			if(baseAttributes.Category() != "Intrasolar" && (!hyperDrive && !jumpDrive))
				checks.emplace_back("no hyperdrive?");
			if(fuelCapacity < navigation.JumpFuel())
				checks.emplace_back("no fuel?");
		}
```

Zitchas commented:
Good idea! With the larger Sol system (and I hope that eventually we will have many such larger systems) it makes sense that there would be intrasolar ships. I think there's even a faction or two that would probably make considerable use of such an attribute.

I do think this would be better as an attribute rather than a category, though. I could easily see people wanting to strip the HD off a Flivver to do a Luna-to-Pluto race. Having the attribute also suppress the bay warning would also be good. I've oft thought that it'd make sense for, say, local defense missions to fly a fighter with a bunch of fighter or drone escorts rather than a larger ship. 

As an aside, categories are clunky and subjective, since only one can be applied to a ship and many ships can appear to qualify for more than one category. I'd personally like to see categories dispensed with in favor of having descriptive tags the player can choose to sort by, or sorting by attributes such as price, mass, weapon space, health, etc. No work has been done on doing such a change, but I suspect it will come at some point. As such, better to put code & focus on attributes rather than categories.

MadisonianX commented:
Updated to include an `intrasolar` attribute for ships or outfits that will exempt it from the Hyperdrive/Jump Drive flight checks, and if on a carried ship Category like Fighter, also exempt it from the Insufficient Bays warning. 

Added an intrasolar attribute check to this line in PlayerInfo:
`if(ship->GetSystem() && !ship->IsDisabled() && !ship->IsParked() && !ship->Attributes().Get("intrasolar"))
`

It seems to work but unsure if the logic is the best location for it or not. 

## Additional testing: 
Added `intrasolar 1` to a Hogshead and purchased one, then sold the Hyperdrive. Was able to take off with no Hyperdrive warning.

Added `intrasolar 1` to a Dagger fighter and purchased one. Was able to take off with no Hyperdrive warning or Insufficient Bays warning.

Alternatively, added `intrasolar 1` to "X1050 Ion Engines" and purchased a Barb fighter which uses them. Same result, was able to take off without warnings.

MadisonianX commented:
If ship categories are eventually deprecated, so be it. Until then I think it makes sense to also have the Intrasolar category as it would specifically be for ships/variants of any size that are stripped of fuel capacity and hyperdrives—possibly even weapons systems/mounts—to make more space available.

Zitchas commented:
Sure! I have no objection, and in fact it does make sense that if there are ships specifically designed to be intrasolar, that they are categorized as such. Having that plus the attribute is good.

TheGiraffe3 commented on data/categories.txt:
```diff
@@ -29,6 +29,7 @@ category "ship"
 	"Unclassified"
 	"Unclassified Minor"
 	"Game Mechanics"
+	"Intrasolar"
```
Is this category still needed as it's not used?

MadisonianX replied:
If the code for both intrasolar category and attribute are merged, I think it'd be better to have the category already defined so it can be documented (with comment in categories.txt and/or wiki) and in-place for usage rather than just something hidden in the code, even if there are presently no ships or variants defined.

Zitchas replied:
Yes, having the category defined in anticipation of having ships to go in it is OK.

Zitchas commented:
The checks are failing, as best I can tell, because it touched ship conditions, so it ran the integration tests, and is failing due to pylons, not because of anything your code does. I'll have to fix that part. (in a stand-alone PR)

Zitchas commented:
I'm expecting the same failure rate as before, but to have fewer errors in the log. That'll show that I'm on the wrong track to fixing the tests. This shouldn't affect your work on the PR itself.

Zitchas commented:
OK, I'm hopeful this time that most of the errors in those test should be gone, possibly even all of them. 

edit: OK, lots fewer errors, but still something cropping up in the assertions test. It'll take me some more time to figure out why it's failing. But in any case, it doesn't appear to concern your code. (at least, not directly. It *might* be failing because ships now effectively have the new attribute with a 0 value, and it's testing isn't taking the existence of it into account, but that's not on your code, it's on the test.

Zitchas commented on source/Ship.cpp:
```diff
@@ -1392,13 +1393,20 @@ vector<string> Ship::FlightCheck() const
 		if(fuel < 0.)
 			checks.emplace_back("fuel?");
-		if(!canBeCarried)
+		if(canBeCarried && !intrasolar)
```
```suggestion
		if(!canBeCarried && !intrasolar)
```

This code section originally applied these two checks to ships that cannot be carried. Meaning ships designed for HD travel. As such, I'm curious as to the logic of switching this check to ships that can be carried and don't have the intrasolar attribute. 

Possible combinations (all ships will be one of these 4 situations):
canBeCarried + Intrasolar = no fuel or HD checks
!canBeCarried + Intrasolar = no fuel or HD checks
canBeCarried + !Intrasolar = no fuel or HD checks
!canBeCarried + !Intrasolar = Do the fuel and HD checks

As such, the correct situation for this if statement to be true and do the HD and fuel checks would be when the ship has both !canBeCarried and !Intrasolar.

Zitchas commented on source/Ship.cpp:
```diff
@@ -1392,13 +1393,20 @@ vector<string> Ship::FlightCheck() const
 		{
 			if(!hyperDrive && !jumpDrive)
 				checks.emplace_back("no hyperdrive?");
 			if(fuelCapacity < navigation.JumpFuel())
 				checks.emplace_back("no fuel?");
```
```suggestion
```

I am thinking that it'd be best to just remove the fuel check from here, so that it can instead be a stand-alone check that triggers on the presence of a means of jumping instead of being based on whether or not it can be carried.

Zitchas commented on source/Ship.cpp:
```diff
@@ -1392,13 +1393,20 @@ vector<string> Ship::FlightCheck() const
 			if(fuelCapacity < navigation.JumpFuel())
 				checks.emplace_back("no fuel?");
 		}
+		if(!canBeCarried)
+		{
+			if(baseAttributes.Category() != "Intrasolar" && (!hyperDrive && !jumpDrive))
+				checks.emplace_back("no hyperdrive?");
+			if(fuelCapacity < navigation.JumpFuel())
+				checks.emplace_back("no fuel?");
+		}
```
```suggestion
		if(hyperDrive || jumpDrive)
		{
			if(fuelCapacity < navigation.JumpFuel())
				checks.emplace_back("no fuel?");
		}
```

This might be more to the point, skipping asking whether or not the ship can be carried or is intrasolar or anything, and just ask "does it have a drive?" If it does, check fuel. If not, don't.

Zitchas reviewed (changes requested):
Thinking about this, there's really two separate things here that have been lumped together:

1) Is this ship designed to be capable of independently leaving the system?

2) Is this ship equipped to leave the system?

Previously, both were lumped into the simple "Is this ship not designed to be carried in a bay? If if it not, then check for both HD and fuel. 

But we can put HD and fuel on carried ships and fly them around. Likewise we could with intrasolar ships, too. 

So I am thinking that it may be better to spit the two checks. So the first check is looking at whether or not the ship is generally intended for interstellar flight. (not designed to be carried, and not intrasolar). If this is the case, then check for HD or Jump drives.

Then the second check is "if the ship has a means of doing a hump, then check if it has the fuel to make a jump."


edit: Looking through the test code, and it occured to me that the fact that this is changing how the checks happen, which *may* result in the tests triggering one of these warning messages (some of the ships in the test don't have HD) differently than the check expects. Getting a warning dialog when the test doesn't expect one *may* break the test. Just supposition, though.

MadisonianX reviewed (commented):
Logic looks good and streamlined, just have to add a ship category check too.

Zitchas commented:
Looks good at a glance, but I would like to test it first.

Zitchas commented:
Just tested it, and it acts properly.

One thought, though: Perhaps we should retain the warning logo on the ship icon (since identifying that a ship isn't HD capable is generally important, since most of the game *is* reliant on being HD capable), and just have the intrasolar attribute prevent the popup warning?

Zitchas commented:
A further thought: I think merging this as-is will be good, but I think that refining the warning/error imagery on the ship icons would be worth a follow-up PR.

Just thinking that currently we just have two icons that go on ship icons in the outfitter/shipyard side bar: error, and warning. The yellow one for "problematic but not major" and the red for "major problem." But that's it. I'm thinking that possibly we should have an icon specifically for "lacks means of jumping", which would be yellow for intrasolar ships, and red for non-intrasolar ships. For intrasolar ships, lacking an HD is commonplace. It's not a blocking thing, but it is worth letting the player know that it is the case. Especially since the player *can* put an HD on an intrasolar ship to make it interstellar, then remove it again later. (say, moving from one big system to another). But on a non-intrasolar ship, not having a means of jumping is probably something really important that the player needs to pay attention to.
